File: aardio/lib/process/ruby/.res/process.rpc.rb
require 'json'
require 'open3'

source_files = ENV['RUBY_SERVER_SOURCE']
unless (source_files||"").empty?
  source_files.split(':').each do |file|
    require_relative file
  end
end

def do_rpc(rpc)
  params = rpc[:params] 
  method = rpc[:method].to_sym

  begin   
    if params.is_a?(Array)
      result = send(method, *params) 
    elsif params.is_a?(Hash)
      result = send(method, **params) 
    else
      raise ArgumentError, "Invalid params type"
    end
      
    rpc_result = {
      jsonrpc: "2.0",
      result: result,
      id: rpc[:id]
    }
     
  rescue StandardError => e
    rpc_result = {
      jsonrpc: "2.0",
      error: {
        code: -32601,
        message: "Procedure not found.",
        data: e.to_s
      },
      id: rpc[:id]
    }
  end

  JSON.generate(rpc_result) + "\n"
end

stdin = IO.new(0)
loop do
  input = stdin.gets
  break if input.nil?

  begin
    json_data = JSON.parse(input, symbolize_names: true)
    
    output = do_rpc(json_data)
    $stdout.puts output
    $stdout.flush
  rescue JSON::ParserError
    print '{"jsonrpc": "2.0", "error": {"code": -32700, "message": "Parse error"}, "id": null}'
  end
end

File: aardio/lib/process/r/.res/process.rpc.r
?>
library( "jsonlite" )
library( "utils" )

source_files <- Sys.getenv( "R_SERVER_SOURCE" )
if( source_files != "" ) {
	source_files <- strsplit( source_files, ":" )[[1]]
	for( s in source_files )
		source( s )
}

do.rpc <- function( rpc )
{
	rpc$params <- as.list( rpc$params )

	result <- try( do.call( rpc$method, rpc$params ), silent = TRUE )

	if( class( result ) == "try-error" ) {
		rpc_result <- list(
				jsonrpc = unbox("2.0"),
				error = list( code = -32601, message = "Procedure not found.", data = as.character( result ) ),
				id = unbox(rpc$id)
				)
	} else {
		rpc_result <- list(
				jsonrpc = unbox("2.0"),
				result = result,
				id = unbox(rpc$id)
				)
	}

	ret <- toJSON( rpc_result,auto_unbox=<? 
  	= owner.autoUnbox ? "TRUE" : "FALSE" ?>,null="null",pretty=FALSE )
	ret <- paste( ret, "\n", sep="" )
	return( ret )
}

process_stdin <- file("stdin",blocking=TRUE,open="rb")
  
while (TRUE) {
  s <- readLines(process_stdin, n=1)

  if (length(s)==0)
    break

  #禁止单元素数组转原子向量(避免 auto_unbox)
  json_data <- try(fromJSON(s,simplifyVector=<? 
  	= owner.simplifyVector ? "TRUE" : "FALSE" ?>,simplifyDataFrame=<? 
  	= owner.simplifyDataFrame ? "TRUE" : "FALSE" ?>,simplifyMatrix=<? 
  	= owner.simplifyMatrix ? "TRUE" : "FALSE" ?>),silent=TRUE)

  if (class(json_data) == "try-error") {
    cat('{"jsonrpc": "2.0", "error": {"code": -32700, "message": "Parse error"}, "id": null}')
  } else {
    ret <- do.rpc(json_data)
    cat(ret)
  }
}

q()
